### PR TITLE
Feature/cxspa 262

### PR DESCRIPTION
**Pull request by WeizhengSap**
enable strict mode for `storefrontlib/cms-components`

### CODE REVIEW

File: projects/core/src/model/cms.model.ts [hunk 1]
```diff
@@ -176,6 +176,7 @@ export interface CmsNavigationComponent extends CmsComponent {
   notice?: string;
   showLanguageCurrency?: string;
   navigationNode?: CmsNavigationNode;
+  resetMenuOnClose?: boolean;
```
Comment (plabadie) [inline projects/core/src/model/cms.model.ts:179]: Did you add a whole new attribute?

Comment (WeizhengSap) [reply projects/core/src/model/cms.model.ts:179]: yes, this is required

File: projects/storefrontlib/cms-components/anonymous-consent-management/banner/anonymous-consent-management-banner.component.ts [hunk 2]
```diff
@@ -25,7 +25,7 @@ export class AnonymousConsentManagementBannerComponent implements OnDestroy {
     this.hideBanner();
     const dialog = this.launchDialogService.openDialog(
       LAUNCH_CALLER.ANONYMOUS_CONSENT,
-      null,
+      undefined,
```
Comment (plabadie) [inline projects/storefrontlib/cms-components/anonymous-consent-management/banner/anonymous-consent-management-banner.component.ts:28]: why change the value?

Comment (WeizhengSap) [reply projects/storefrontlib/cms-components/anonymous-consent-management/banner/anonymous-consent-management-banner.component.ts:28]: the second param of `openDialog` is `ElementRef` or `undefined`

Comment (plabadie) [reply projects/storefrontlib/cms-components/anonymous-consent-management/banner/anonymous-consent-management-banner.component.ts:28]: Got it, thanks.

File: projects/core/src/model/cms.model.ts (hunk 1 above)
Comment (plabadie) [reply projects/core/src/model/cms.model.ts:179]: ok.

File: projects/storefrontlib/cms-components/misc/site-context-selector/language-currency.component.ts [hunk 3]
```diff
@@ -1,11 +1,18 @@
...
+    ></cx-site-context-selector>
+    <cx-site-context-selector
+      [context]="SiteContextType.CURRENCY"
+    ></cx-site-context-selector>
   `,
   changeDetection: ChangeDetectionStrategy.OnPush,
 })
-export class LanguageCurrencyComponent {}
+export class LanguageCurrencyComponent {
+  readonly SiteContextType = SiteContextType;
```
Comment (plabadie) [inline projects/storefrontlib/cms-components/misc/site-context-selector/language-currency.component.ts]: What's the new attribute for?
It also probably needs to start with a lowercase letter.

File: projects/storefrontlib/cms-components/misc/site-context-selector/site-context-component.service.ts [hunk 4]
```diff
@@ -71,17 +73,28 @@ export class SiteContextComponentService {
...
+  protected getContext(
+    context?: SiteContextType
+  ): Observable<string | undefined> {
     if (context) {
-      return of(context);
+      if (context === SiteContextType.CURRENCY) {
+        return of(CURRENCY_CONTEXT_ID);
+      } else if (context === SiteContextType.LANGUAGE) {
+        return of(LANGUAGE_CONTEXT_ID);
+      } else {
```
Comment (plabadie) [inline projects/storefrontlib/cms-components/misc/site-context-selector/site-context-component.service.ts:92]: Was that new logic required for strict typing?

File: projects/storefrontlib/cms-components/myaccount/address-book/address-form/suggested-addresses-dialog/suggested-addresses-dialog.component.html [hunk 5]
```diff
@@ -43,7 +43,9 @@
             [for]="'suggested-addresses--suggested-' + i"
           >
             {{ 'addressSuggestion.suggestedAddress' | cxTranslate }}
-            {{ suggestedAddresses?.length > 1 ? i + 1 : null }}
+            {{
+              suggestedAddresses && suggestedAddresses.length > 1 ? i + 1 : null
```
Comment (plabadie) [inline projects/storefrontlib/cms-components/myaccount/address-book/address-form/suggested-addresses-dialog/suggested-addresses-dialog.component.html]: What is the result when `suggestedAddresses` is falsy?

Should we use a default value instead? like 
```
(suggestedAddresses?.length ?? 0) > 1
```

Comment (plabadie) [reply projects/storefrontlib/cms-components/myaccount/address-book/address-form/suggested-addresses-dialog/suggested-addresses-dialog.component.html]: Also, I'm not sure about the reult with operator precedence.  The `>` is executed before the `&&` I think.

File: projects/storefrontlib/cms-components/myaccount/my-interests/my-interests.component.ts [hunk 6]
```diff
@@ -94,10 +94,7 @@ export class MyInterestsComponent implements OnInit, OnDestroy {
     this.sortLabels = this.getSortLabels();
   }
 
-  private getSortLabels(): Observable<{
-    byNameAsc: string;
-    byNameDesc: string;
-  }> {
+  private getSortLabels(): Observable<{ [key: string]: string }> {
```
Comment (plabadie) [inline projects/storefrontlib/cms-components/myaccount/my-interests/my-interests.component.ts]: What's the purpose of this change?  (the method is prvate, but it has impact on a public property)

File: projects/storefrontlib/cms-components/myaccount/payment-methods/payment-methods.component.ts [hunk 7]
```diff
@@ -78,12 +78,12 @@ export class PaymentMethodsComponent implements OnInit {
           actions.push({ name: textDelete, event: 'edit' });
           const card: Card = {
             role: 'region',
-            header: defaultPayment ? textDefaultPaymentMethod : null,
+            header: defaultPayment ? textDefaultPaymentMethod : undefined,
```
Comment (plabadie) [inline projects/storefrontlib/cms-components/myaccount/payment-methods/payment-methods.component.ts:81]: In general (it's done in many places) What id the purpose of replacing `null` by `undefined` ?

File: projects/storefrontlib/cms-components/misc/site-context-selector/language-currency.component.ts (hunk 3 above)
Comment (WeizhengSap) [reply projects/storefrontlib/cms-components/misc/site-context-selector/language-currency.component.ts]: `SiteContextType` is enum, the `[context]` passed to `selector` should be enum. I will rename it lowercase.

File: projects/storefrontlib/cms-components/misc/site-context-selector/site-context-component.service.ts (hunk 4 above)
Comment (WeizhengSap) [reply projects/storefrontlib/cms-components/misc/site-context-selector/site-context-component.service.ts:92]: yes, before it returns `of(context)`, which is `Observable<SiteContextType>`. This function needs return `Observable<string>`.

File: projects/storefrontlib/cms-components/myaccount/address-book/address-form/suggested-addresses-dialog/suggested-addresses-dialog.component.html (hunk 5 above)
Comment (WeizhengSap) [reply projects/storefrontlib/cms-components/myaccount/address-book/address-form/suggested-addresses-dialog/suggested-addresses-dialog.component.html]: I revert it back, `suggestedAddresses` is an array.

File: projects/storefrontlib/cms-components/myaccount/my-interests/my-interests.component.ts (hunk 6 above)
Comment (WeizhengSap) [reply projects/storefrontlib/cms-components/myaccount/my-interests/my-interests.component.ts]: revert it back

File: projects/storefrontlib/cms-components/myaccount/payment-methods/payment-methods.component.ts (hunk 7 above)
Comment (WeizhengSap) [reply projects/storefrontlib/cms-components/myaccount/payment-methods/payment-methods.component.ts:81]: because `Card's header` is `header?: string;`. So, it can be either `string` or `undefined`

File: projects/storefrontlib/cms-components/myaccount/payment-methods/payment-methods.component.ts [hunk 8]
```diff
@@ -96,24 +96,28 @@ export class PaymentMethodsComponent implements OnInit {
   }
 
   deletePaymentMethod(paymentMethod: PaymentDetails): void {
-    this.userPaymentService.deletePaymentMethod(paymentMethod.id);
-    this.editCard = null;
+    if (paymentMethod.id) {
+      this.userPaymentService.deletePaymentMethod(paymentMethod.id);
+      this.editCard = undefined;
```
Comment (plabadie) [inline projects/storefrontlib/cms-components/myaccount/payment-methods/payment-methods.component.ts:101]: Looks like `this.editCard = undefined;` should be outside of the `if`

File: projects/storefrontlib/cms-components/myaccount/payment-methods/payment-methods.component.ts [hunk 9]
```diff
@@ -96,24 +96,28 @@ export class PaymentMethodsComponent implements OnInit {
...
-      { key: 'paymentMessages.setAsDefaultSuccessfully' },
-      GlobalMessageType.MSG_TYPE_CONFIRMATION
-    );
+    if (paymentMethod.id) {
+      this.userPaymentService.setPaymentMethodAsDefault(paymentMethod.id);
+      this.globalMessageService?.add(
+        { key: 'paymentMessages.setAsDefaultSuccessfully' },
+        GlobalMessageType.MSG_TYPE_CONFIRMATION
+      );
+    }
```
Comment (plabadie) [inline projects/storefrontlib/cms-components/myaccount/payment-methods/payment-methods.component.ts]: I think that's probably ok because the payment method id should probably not ever be falsy if this function is called.
I was wondering however if we should throw an exception or something like this instead of silently doing nothing.  Let's discuss.

File: projects/storefrontlib/cms-components/navigation/category-navigation/category-navigation.component.html [hunk 10]
```diff
@@ -1,8 +1,8 @@
 <cx-navigation-ui
   *ngIf="data$ | async as data"
   [node]="node$ | async"
-  [ngClass]="data.styleClass"
-  [wrapAfter]="data.wrapAfter"
+  [ngClass]="data.styleClass ?? ''"
+  [wrapAfter]="+(data.wrapAfter ?? '')"
```
Comment (plabadie) [inline projects/storefrontlib/cms-components/navigation/category-navigation/category-navigation.component.html:5]: What's the `+` for, converting the value to a string?

File: projects/storefrontlib/cms-components/navigation/search-box/search-box.component.ts [hunk 11]
```diff
@@ -229,7 +230,7 @@ export class SearchBoxComponent implements OnInit, OnDestroy {
   }
 
   // Focus on next item in results list
-  focusNextChild(event: UIEvent) {
+  focusNextChild(event: UIEvent | Event) {
```
Comment (plabadie) [inline projects/storefrontlib/cms-components/navigation/search-box/search-box.component.ts]: If the param needs to be type `Event` shouldn't we delete `UIEvent` from the signature since `UIEvent` extends `Event` anyways ?

Same goes for the other similar changed in this file.

(Or is it not the same Event type?)

File: projects/storefrontlib/cms-components/product/product-list/product-facet-navigation/active-facets/active-facets.component.html [hunk 12]
```diff
@@ -1,13 +1,13 @@
 <ng-container *ngIf="facetList$ | async as facetList">
-  <h4 *ngIf="facetList?.activeFacets?.length > 0">
+  <h4 *ngIf="facetList.activeFacets && facetList.activeFacets.length > 0">
```
Comment (plabadie) [inline projects/storefrontlib/cms-components/product/product-list/product-facet-navigation/active-facets/active-facets.component.html:2]: Still not sure about operator precedence between && and >.
Do you think it could be an issue? Perhaps we need to try it out.

File: projects/storefrontlib/cms-components/product/product-tabs/product-reviews/product-reviews.component.ts [hunk 13]
```diff
@@ -99,7 +105,9 @@ export class ProductReviewsComponent {
       alias: reviewFormControls.reviewerName.value,
     };
 
-    this.reviewService.add(product.code, review);
+    if (product.code) {
```
Comment (plabadie) [inline projects/storefrontlib/cms-components/product/product-tabs/product-reviews/product-reviews.component.ts]: Should we throw an error of some kind or just silently not call the function?

File: projects/storefrontlib/shared/components/list-navigation/pagination/pagination.component.ts [hunk 14]
```diff
@@ -37,9 +37,11 @@ export class PaginationComponent {
   get pagination(): PaginationModel {
     return this._pagination;
   }
-  @Input() set pagination(value: PaginationModel) {
-    this._pagination = value;
-    this.render(value);
+  @Input() set pagination(value: PaginationModel | undefined) {
```
Comment (plabadie) [inline projects/storefrontlib/shared/components/list-navigation/pagination/pagination.component.ts:40]: Is there a way to not accept `undefined` if we are not goint to do anything when we recieve it?

File: projects/storefrontlib/cms-components/myaccount/payment-methods/payment-methods.component.ts (hunk 8 above)
Comment (WeizhengSap) [reply projects/storefrontlib/cms-components/myaccount/payment-methods/payment-methods.component.ts:101]: only payment method is deleted, then this `editCard` can be undefined.

File: projects/storefrontlib/cms-components/myaccount/payment-methods/payment-methods.component.ts (hunk 9 above)
Comment (WeizhengSap) [reply projects/storefrontlib/cms-components/myaccount/payment-methods/payment-methods.component.ts:120]: If there is "PaymentDetails", it must have id. (like Product, if there is product, it must have code). But, we defined `id` as optional. So, customers can call it `setDefaultPaymentMethod({})`.

File: projects/storefrontlib/cms-components/navigation/search-box/search-box.component.ts (hunk 11 above)
Comment (WeizhengSap) [reply projects/storefrontlib/cms-components/navigation/search-box/search-box.component.ts]: I forgot to remove `| Event`.

File: projects/storefrontlib/cms-components/navigation/category-navigation/category-navigation.component.html (hunk 10 above)
Comment (WeizhengSap) [reply projects/storefrontlib/cms-components/navigation/category-navigation/category-navigation.component.html:5]: convert string to number

File: projects/storefrontlib/cms-components/product/product-tabs/product-reviews/product-reviews.component.ts (hunk 13 above)
Comment (WeizhengSap) [reply projects/storefrontlib/cms-components/product/product-tabs/product-reviews/product-reviews.component.ts]: if there is product, it must have code. I can see that in most cases, we just use default ` ?? ''` for it. I can do it as well.

File: projects/storefrontlib/shared/components/list-navigation/pagination/pagination.component.ts (hunk 14 above)
Comment (WeizhengSap) [reply projects/storefrontlib/shared/components/list-navigation/pagination/pagination.component.ts:40]: `pagination` defined in interfaces are `optional`. The internal one is this `_pagination: PaginationModel`. So, if `undefined` is passed, it just not do rendering

File: projects/storefrontlib/cms-components/product/product-list/product-facet-navigation/active-facets/active-facets.component.html (hunk 12 above)
Comment (WeizhengSap) [reply projects/storefrontlib/cms-components/product/product-list/product-facet-navigation/active-facets/active-facets.component.html:2]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence.
So, `>` is higher than `&&`

**Review by plabadie — APPROVED** [review]

**Review by plabadie — CHANGES_REQUESTED** [review]
There is a unit test failing

**Review by plabadie — APPROVED** [review]
